x/concentrated-liquidity/position.go

```diff
@@ -226,3 +228,101 @@ func (k Keeper) getNextPositionIdAndIncrement(ctx sdk.Context) uint64 {
+	if len(positionIds) <= MinNumPositionsToCombine {
+		return 0, types.PositionQuantityTooLowError{MinNumPositions: MinNumPositionsToCombine, NumPositions: len(positionIds)}
+	}
+
+	// Check that all the positions are in the same pool, tick range, and are fully charged.
+	// Sum the liquidity of all the positions.
+	poolId, lowerTick, upperTick, liquidity, err := k.validatePositionsAndGetTotalLiquidity(ctx, owner, positionIds)
+	if err != nil {
+		return 0, err
+	}
+
+	// The new position's timestamp is the current block time minus the fully charged duration.
+	joinTime := ctx.BlockTime().Add(-types.FullyChargedDuration)
+
+	// Get the next position ID and increment the global counter.
+	positionId := k.getNextPositionIdAndIncrement(ctx)
+
+	// Initialize the fee accumulator for the new position.
+	if err := k.initializeFeeAccumulatorPosition(ctx, poolId, lowerTick, upperTick, positionId); err != nil {
+		return 0, err
+	}
+
+	// Check if the position already exists.
+	hasFullPosition := k.hasFullPosition(ctx, positionId)
+	if !hasFullPosition {
+		// If the position does not exist, initialize it with the provided liquidity and tick range.
+		err = k.initOrUpdatePositionUptime(ctx, poolId, liquidity, owner, lowerTick, upperTick, sdk.ZeroDec(), joinTime, positionId)
+		if err != nil {
+			return 0, err
+		}
+	} else {
+		// If the position already exists, return an error.
+		return 0, err
+	}
+
+	// Update the position in the pool based on the provided tick range and liquidity delta.
+	_, _, err = k.updatePosition(ctx, poolId, owner, lowerTick, upperTick, liquidity, joinTime, positionId)
+	if err != nil {
+		return 0, err
+	}
+
+	// Delete the previous positions.
+	for _, positionId := range positionIds {
+		err := k.deletePosition(ctx, positionId, owner, poolId)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	return positionId, nil
+}
+
+// validatePositionsAndGetTotalLiquidity checks that the positions are all in the same pool and tick range, and returns the total liquidity of the positions.
+func (k Keeper) validatePositionsAndGetTotalLiquidity(ctx sdk.Context, owner sdk.AccAddress, positionIds []uint64) (uint64, int64, int64, sdk.Dec, error) {
+	totalLiquidity := sdk.ZeroDec()
+	// Note the first position's params to use as the base for comparison.
+	basePosition, err := k.GetPosition(ctx, positionIds[0])
+	if err != nil {
+		return 0, 0, 0, sdk.Dec{}, err
+	}
+
+	for i, positionId := range positionIds {
+		position, err := k.GetPosition(ctx, positionId)
+		if err != nil {
+			return 0, 0, 0, sdk.Dec{}, err
+		}
+		// Check that the caller owns all the positions.
+		if position.Address != owner.String() {
+			return 0, 0, 0, sdk.Dec{}, types.PositionOwnerMismatchError{PositionOwner: position.Address, Sender: owner.String()}
+		}
+
+		// Check that all the positions are fully charged.
+		fullyChargedMinTimestamp := position.JoinTime.Add(types.FullyChargedDuration)
+		if fullyChargedMinTimestamp.After(ctx.BlockTime()) {
```

Comment:
Nit: this should be `!fullyChargedMinTimestamp.Before(ctx.BlockTime())` since we would want positions that are charged for exactly two weeks to qualify (e.g. if someone wants to chain fungifications together). Up to you to make this change or leave as is

Reply:
https://github.com/osmosis-labs/osmosis/pull/4800/commits/6568718072b72804aa07e59862799455398883c2